ivadomed/postprocessing.py

```diff
@@ -0,0 +1,205 @@
+
+import numpy as np
+import nibabel as nib
+from scipy.ndimage.measurements import label
+from scipy.ndimage.morphology import binary_fill_holes
+
+
+def threshold_predictions(predictions, thr=0.5):
+    """Threshold predictions.
+
+    Threshold a soft (ie not binary) array of predictions given a threshold value, and returns
+    a binary array.
+
+    Args:
+        predictions (array): array to binarise.
+        thr (float): Threshold value: voxels with a value < to thr are assigned 0 as value, 1
+            otherwise.
+    Returns:
+        array: Array containing only zeros or ones.
+
+    """
+    thresholded_preds = np.copy(predictions)[:]
+    low_values_indices = thresholded_preds < thr
+    thresholded_preds[low_values_indices] = 0
+    low_values_indices = thresholded_preds >= thr
+    thresholded_preds[low_values_indices] = 1
+    return thresholded_preds.astype(np.int)
+
+
+def threshold_predictions_nib(nib_predictions, thr=0.5):
+    """Threshold predictions from a nibabel object.
+
+    Threshold a soft (ie not binary) array of predictions from a nibabel object given a threshold value,
+    and returns a nibabel object with binary data.
+    Note: this function calls threshold_predictions().
+
+    Args:
+        nib_predictions (nibabelObject): nibabel image to binarise.
+        thr (float): Threshold value: voxels with a value < to thr are assigned 0 as value, 1
+            otherwise.
+    Returns:
+        nibabelObject: Nibabel object containing only zeros or ones.
+
+    """
+    data = nib_predictions.get_fdata()
+    data_thr = threshold_predictions(predictions=data, thr=thr)
+    return nib.Nifti1Image(data_thr, nib_predictions.affine)
+
+
+def keep_largest_object(predictions):
+    """Keep the largest connect object.
+
+    Keep the largest connected object from the input array (2 or 3D).
+    Note: This function only works for binary segmentation.
+
+    Args:
+        predictions (array): Input 2 or 3D binary segmentation.
+    Returns:
+        array: processed segmentation.
+
+    """
+    assert predictions.dtype == np.dtype('int')
+    # Find number of closed objects using skimage "label"
+    labeled_obj, num_obj = label(predictions)
+    # If more than one object is found, keep the largest one
+    if num_obj > 1:
+        # Keep the largest object
+            predictions[np.where(labeled_obj != (np.bincount(labeled_obj.flat)[1:].argmax() + 1))] = 0
+    return predictions
+
+
+def keep_largest_object_nib(nib_predictions):
+    """Keep the largest connect object from nibabel image.
+
+    Keep the largest connected object from the input image (2 or 3D).
+    Note: This function only works for binary segmentation.
+    Note: This function calls keep_largest_object().
+
+    Args:
+        nib_predictions (nibabelObject): Input image, 2 or 3D, binary segmentation.
+    Returns:
+        nibabelObject: processed segmentation.
+
+    """
+    data = nib_predictions.get_fdata()
```

Comment:
here we need to be extra careful with this new `.get_fdata()` method, which by default opens an image as float64, see [here](https://nipy.org/nibabel/reference/nibabel.dataobj_images.html#nibabel.dataobj_images.DataobjImage.get_fdata).

As a consequence, even if an int nibabel object is passed, getting the data at this line will transform it to a float64, and the code will crash here: 
https://github.com/neuropoly/ivado-medical-imaging/pull/184/files#diff-632badec6c3b73f6da0d616d49a5bea8R104

Reply:
@kousu any suggestion?

Reply:
I think `nib_predictions.header.get_data_dtype()`, and [`np.asanyarray(nib_predictions.dataobj)`](https://nipy.org/nibabel/reference/nibabel.dataobj_images.html#nibabel.dataobj_images.DataobjImage.get_data) which must be implemented with the help of [`nib_predictions.dataobj.dtype`](https://nipy.org/nibabel/reference/nibabel.arrayproxy.html#nibabel.arrayproxy.ArrayProxy.dtype).

I can see why they wanted to make the output type more predictable, but forcing float64 on everyone is not helpful for us, or for a lot of formats that know their types just fine.

I think we probably want to replace the deprecated `.get_data()` call everywhere with `.dataobj.get_unscaled()`, to work around their helpfulness. But maybe I'm missing some context for that.

Reply:
> I think we probably want to replace the deprecated .get_data() call everywhere with .dataobj.get_unscaled(), to work around their helpfulness. But maybe I'm missing some context for that.

right now we've replaced all these calles with `get_fdata()` (as per nibabel's team suggestion), but your suggestion might be better. We should probably dedicate a specific issue to sort this out. Anyone welcome to look into it.

Reply:
> ```
> data = nib_predictions.dataoobj.get_unscaled()
> ```

unfortunately this does not work as nib are `Nifti1Image` object, which does not have the field `dataoobj` (it has `dataobj`, but this one does not have the `get_unscaled()` method).

Reply:
Drat. `im.get_fdata(dtype=im.header.get_data_dtype())`?

Reply:
I think your other suggestion is fine: `np.asanyarray(nii_im.dataobj)` so i went ahead with it

Reply:
@kousu i am running into a "shared" address issue:
```python
data = np.array([0, 1, 2])
nii = nib.nifti1.Nifti1Image(data, np.eye(4))
nii.dataobj
# >> array([0, 1, 2])
nii_new = nib.nifti1.Nifti1Image(np.asanyarray(nii.dataobj), np.eye(4))
nii_new.dataobj[1] = 0
nii.dataobj
# >> array([0, 0, 2]) --> # I EXPECT THIS TO BE array([0, 1, 2])
```

Reply:
ah! this is solved by replacing 
````
nii_new = nib.nifti1.Nifti1Image(np.asanyarray(nii.dataobj), np.eye(4))
````
with
````
nii_new = nib.nifti1.Nifti1Image(np.copy(np.asanyarray(nii.dataobj)), np.eye(4))
````

Reply:
Well, this is on purpose to avoid unnecessary copies [in numpy](https://docs.scipy.org/doc/numpy/reference/generated/numpy.asanyarray.html):

>     If a is an ndarray or a subclass of ndarray, it is returned as-is and no copy is performed.

In the spirit of "explicit is better than implicit", you can write

```
nii_new = nib.nifti1.Nifti1Image(np.copy(np.asanyarray(nii.dataobj)), np.eye(4))
```

but something more bugs me about this. I'm experimenting.

Reply:
There's some bizarre caching going on inside of `DataObjImage.get_fdata()`. I haven't fully understood it yet. It seems that it may or may not hand back the same object depending on whether it's been edited in the meantime. I would expect it to either make a new copy *each time*, but it doesn't; it seems to make a copy *per dtype*, and one copy per edit made to a previously returned array.

By the way, `.get_data()` always hands back the same pointer (`im.get_data() is im.dataobj == True`), so you would have the same shared-pointer problem; unless they changed that when they introduced `.get_fdata()`.

<details>
```
COLORS = ['🍵', '🍏', '🍭', '🍄', '🥕', '🍇', '🥒', '🍅',  '🍊']
IDS = {}

def colorid(obj):
    "rewrite pointers as emojis, for easier visual inspection"
    if id(obj) not in IDS:
        IDS[id(obj)] = COLORS.pop()
    return IDS[id(obj)]
    
    
def p(var): # debugging hack
    value = eval(var)
    print("%35s = %40r @ %s" % (var, value, colorid(value)))

import numpy as np
import nibabel as nib

data = np.array([0, 1, 2])

p('data')

print("Comparing get_data() to get_fdata()")
print()
nii = nib.nifti1.Nifti1Image(data, np.eye(4))
p('nii.dataobj')
p('nii.get_data()')
p('nii.get_fdata()')


print()
print("Extracting data twice in two ways: get_f?data() - to see if we get the same pointers both times")
data = nii.get_data()
fdata = nii.get_fdata()
data2 = nii.get_data()
fdata2 = nii.get_fdata()

p("nii.dataobj")
p('nii.get_data()')
p('data')
p('data2')
print()
p('nii.get_fdata()')
p('fdata')
p('fdata2')

print()
print("Editing fdata")
fdata[2] += 8

p("nii.dataobj")
p('nii.get_data()')
p('data')
p('data2')
print()
p('nii.get_fdata()')
p('fdata')
p('fdata2')


print()
print("Extracting data a third time")
data3 = nii.get_data()
fdata3 = nii.get_fdata()

p("nii.dataobj")
p('nii.get_data()')
p("data3")
print()
p('nii.get_fdata()')
p("fdata3")


fdata_float16 = nii.get_fdata(dtype=np.float16)
fdata_float32 = nii.get_fdata(dtype=np.float32)
p("fdata_float16")
p("fdata_float32")

print()
print("Editing data via fdata_float32")
fdata_float32[1] -= 88


p("nii.dataobj")
p('nii.get_data()')
p('data')
p('data2')
p("data3")
print()
p('fdata')
p('fdata2')
p("fdata3")
p('nii.get_fdata()')
p("fdata_float16")
p('nii.get_fdata(dtype=np.float16)')
p("fdata_float32")
p('nii.get_fdata(dtype=np.float32)')
p('nii.get_fdata(dtype=np.float64)')
```

producing

```
 python test.py 
                               data =                         array([0, 1, 2]) @ 🍊
Comparing get_data() to get_fdata()

                        nii.dataobj =                         array([0, 1, 2]) @ 🍊
                     nii.get_data() =                         array([0, 1, 2]) @ 🍊
                    nii.get_fdata() =                      array([0., 1., 2.]) @ 🍅

Extracting data twice in two ways: get_f?data() - to see if we get the same pointers both times
                        nii.dataobj =                         array([0, 1, 2]) @ 🍊
                     nii.get_data() =                         array([0, 1, 2]) @ 🍊
                               data =                         array([0, 1, 2]) @ 🍊
                              data2 =                         array([0, 1, 2]) @ 🍊

                    nii.get_fdata() =                      array([0., 1., 2.]) @ 🍅
                              fdata =                      array([0., 1., 2.]) @ 🍅
                             fdata2 =                      array([0., 1., 2.]) @ 🍅

Editing fdata
                        nii.dataobj =                         array([0, 1, 2]) @ 🍊
                     nii.get_data() =                         array([0, 1, 2]) @ 🍊
                               data =                         array([0, 1, 2]) @ 🍊
                              data2 =                         array([0, 1, 2]) @ 🍊

                    nii.get_fdata() =                   array([ 0.,  1., 10.]) @ 🍅
                              fdata =                   array([ 0.,  1., 10.]) @ 🍅
                             fdata2 =                   array([ 0.,  1., 10.]) @ 🍅

Extracting data a third time
                        nii.dataobj =                         array([0, 1, 2]) @ 🍊
                     nii.get_data() =                         array([0, 1, 2]) @ 🍊
                              data3 =                         array([0, 1, 2]) @ 🍊

                    nii.get_fdata() =                   array([ 0.,  1., 10.]) @ 🍅
                             fdata3 =                   array([ 0.,  1., 10.]) @ 🍅
                      fdata_float16 =       array([0., 1., 2.], dtype=float16) @ 🥒
                      fdata_float32 =       array([0., 1., 2.], dtype=float32) @ 🍇

Editing data via fdata_float32
                        nii.dataobj =                         array([0, 1, 2]) @ 🍊
                     nii.get_data() =                         array([0, 1, 2]) @ 🍊
                               data =                         array([0, 1, 2]) @ 🍊
                              data2 =                         array([0, 1, 2]) @ 🍊
                              data3 =                         array([0, 1, 2]) @ 🍊

                              fdata =                   array([ 0.,  1., 10.]) @ 🍅
                             fdata2 =                   array([ 0.,  1., 10.]) @ 🍅
                             fdata3 =                   array([ 0.,  1., 10.]) @ 🍅
                    nii.get_fdata() =                      array([0., 1., 2.]) @ 🥕
                      fdata_float16 =       array([0., 1., 2.], dtype=float16) @ 🥒
    nii.get_fdata(dtype=np.float16) =       array([0., 1., 2.], dtype=float16) @ 🍄
                      fdata_float32 = array([  0., -87.,   2.], dtype=float32) @ 🍇
    nii.get_fdata(dtype=np.float32) =       array([0., 1., 2.], dtype=float32) @ 🥕
    nii.get_fdata(dtype=np.float64) =                      array([0., 1., 2.]) @ 🍭
```
</details>